### DISCUSSION TITLE
New Feature: AVPF Configurable per Account

### DISCUSSION THREAD

**Discussion by maximilianfridrich** [baresip/baresip#3225] [General]
I am currently looking into making changes to BareSIP so that the RTP profile `RTP/AVPF` ([RFC 4585](https://www.rfc-editor.org/rfc/rfc4585)) can be configured by the user such that it is the protocol defined in the SDP media line. This is necessary according to [section 4.1 of RFC 4585](https://www.rfc-editor.org/rfc/rfc4585#section-4.1) and there are other UAs which require this:

```
   Feedback information following the modified timing rules as specified
   in this document MUST NOT be sent for a particular media session
   unless the description for this session indicates the use of the
   "AVPF" profile (exclusively or jointly with other AV profiles).
```

For this, I would like to have some input from the community/maintainers:

**How should the RTP profile be configured by the user?**

Currently, other RTP profiles like `RTP/SAVP` are configured by setting `mediaenc={srtp,srtp-mand,srtp-mandf,...}` as an account parameter. Clearly, `AVPF` should then also be configurable per account.

- Should this be a new account option and if so, how should this option be called?
- Or, for simplicity sake, should a new pseudo `mediaenc=avpf` be added?
- Or something else?

Any feedback is appreciated. @alfredh @cspiel1 @sreimers @juha-h

**Comment by cspiel1** [comment]
Because it doesn't really fit to `menc` maybe a flag called:
```
#  addr-params:
#  ;mediaf={yes, no}
```

`f` for feedback

Edit: Should be documented also in the wiki.

**Comment by juha-h** [comment]
How about telling in account config, what media encodings the account accepts and what it proposes in priority order?   The lists could include at least the common ones RTP/AVP, RTP/SAVP, and UDP/TLS/RTP/SAVPF?  This would be clearer that the current mediaenc definition.  In addition, support for ZRTP for needs to be indicated.

**Reply by cspiel1** [reply]
Should we break existing behavior? Or set the current config API to deprecated and after a while drop it?

**Reply by juha-h** [reply]
We don't need to break the existing behavior.  Just introduce new account parameters for accepted and proposed media encodings.  No one needs to use them.

**Reply by juha-h** [reply]
Regarding proposed ones, baresip tries first from the list and if it gets back not accepted here, it tries the next , and so on.  This is how my sip-proxy with rtpengine works.

**Reply by juha-h** [reply]
ZRTP could be separate account config variable `;zrtp=yes/no`, but according to https://datatracker.ietf.org/doc/html/rfc6189, it only applies to RTP/AVP encoding.

**Reply by maximilianfridrich** [reply]
> baresip tries first from the list and if it gets back not accepted here, it tries the next , and so on. This is how my sip-proxy with rtpengine works.

It's a good idea, but I believe it is slightly out of scope right now for my intended feature. For now, I just want to add the option of having `RTP/AVPF` as protocol in the SDP. I see the support of multiple protocol options in the SDP as a valuable but separate feature.

For the `RTP/AVPF` option I would prefer a simpler configuration approach for now, e.g. with the `;mediaf={yes, no}` option or extending `mediaenc={srtp,srtp-mand,srtp-mandf,...}`.

PS:
Regarding multiple protocol options: Usually BareSIP tries an INVITE and if it doesn't work, that's it. Negotiation should happen in the SDP and there are ways to provide multiple transport options within the SDP by combining [RFC 4585](https://www.rfc-editor.org/rfc/rfc4585) with [RFC 3388](https://www.rfc-editor.org/rfc/rfc3388). See [RFC 4585 section 4.4 Example 3](https://www.rfc-editor.org/rfc/rfc4585.html#section-4.4):

```
v=0
      o=alice 3203093520 3203093520 IN IP4 host.example.com
      s=Multicast video with feedback
      t=3203130148 3203137348
      m=audio 49170 RTP/AVP 0
      c=IN IP4 224.2.1.183
      a=rtpmap:0 PCMU/8000
      m=video 51372 RTP/AVP 98 99
      c=IN IP4 224.2.1.184
      a=rtpmap:98 H263-1998/90000
      a=rtpmap:99 H261/90000
      m=video 51372 RTP/AVPF 98 99
      c=IN IP4 224.2.1.184
      a=rtpmap:98 H263-1998/90000
      a=rtpmap:99 H261/90000
      a=rtcp-fb:* nack
      a=rtcp-fb:98 nack rpsi
```

**Reply by juha-h** [reply]
In my opinion,  better to do it in a general way from the beginning, i.e., add rtp_protocols or similar configuration option with possible values RTP/AVP, RTP/AVPF, RTP/SAVP, and UDP/TLS/RTP/SAVPF. Then it would be clear to everyone, what they mean. Regarding negotiation, I suspect that very few SIP UAs implement the multiple transport options feature.

**Reply by juha-h** [reply]
It is not a good idea to always add a new hack when something is missing.

**Reply by maximilianfridrich** [reply]
It's not "from the beginning" though, since `mediaenc` already exists. Your proposed solution adds ambiguity and redundancy to the config, so it's not so clear cut what's the "hacky" solution here and what's the cleaner solution. Let's wait and see if some other people want to chime in.

**Reply by juha-h** [reply]
I already proposed a clearer solution: introduce an account config variable, such as ;rtp_profile, where the offered and accepted rtp profiles are listed. Current possible values in the list are  RTP/AVP, RTP/AVPF, RTP/SAVP, and UDP/TLS/RTP/SAVPF.  In addition, ;zrtp=yes/no would be needed.

But I'll give up.

**Reply by sreimers** [reply]
I also don't think rtp_profile is cleaner solution here. keep in mind, baresip has a modular design. so you could offer:

`rtp_profile=UDP/TLS/RTP/SAVPF` 

but no media encoding module is loaded. So it's not very user friendly to configure the right profiles without knowing much about RFCs and module dependency. And keep in mind, baresip is not only a SIP UA, it can also be used without account config for example within WebRTC context. So moving the configuration from menc module to account config would break many things, as mentioned earlier here.

So I think the feature flag approach is the better solution:
```
#  addr-params:
# ;rtcp_feedback={yes, no} # default yes
```
Then `menc` modules are responsible for checking this flag and baresip core adds `RTP/AVPF` if no menc is set/used.
By default rtcp_feedback should be enabled to keep the current behavior.

**Reply by cspiel1** [reply]
`rtcp_feedback=no` would be the current behavior.

**Reply by sreimers** [reply]
Yes, but depends on menc module:

https://github.com/baresip/baresip/blob/9a4db756b63180f5f25b655073ee2ece96650813/modules/dtls_srtp/dtls_srtp.c#L442-L446

**Comment by maximilianfridrich** [comment]
Thanks @everyone for the feedback. I think we have enough input for now. If we open a PR in the future to implement this, we will reference this discussion.
